### PR TITLE
implement sheet listing all package dependencies from SPM;

### PR DESCRIPTION
**Pull request by NikolaiMadlener**
# *Add an Open-Source Contributions Screen*

## :recycle: Current situation & Problem
The application currently does not provide any way to see all open-source tools used within the app, nor does it give access to all licenses used.
See: [https://github.com/StanfordSpezi/SpeziTemplateApplication/issues/29](url)

## :gear: Release Notes 
- add a info button to the top right corner of Home view that opens the Contributions sheet
- add Contributions sheet that lists all package dependencies used
- package dependencies are read via SwiftPackageList API (using a build phase script, the view is always up do date with the latest dependencies). For more info check out the repo: [Swift-Package-List](https://github.com/FelixHerrmann/swift-package-list)
- each dependency item in the list shows the package name, version/branch, a license label as well as a link to the GitHub repo
- currently detected licenses (shown as tags in the UI) of packages are MIT, Apache-2.0, GPL-2.0, GPL-3.0. Could still be extended in the future.

Here is how it looks like:
Simulator Screenshot - iPhone 14 Pro - 2023-09-18 at 15 42 39 Simulator Screenshot - iPhone 14 Pro - 2023-09-18 at 15 43 13

## :books: Documentation
TBA

## :white_check_mark: Testing
- Simple UI test added that tests the info button and the subsequent existence of the contributions sheet

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by NikolaiMadlener** [comment]
Just a first show for my interpretation of Issue #29.

**Comment by PSchmiedmayer** [comment]
Hi @NikolaiMadlener great to see you here! Let us know once you deem the PR ready for review and tag me if you have any questions! 👍

**Comment by NikolaiMadlener** [comment]
@PSchmiedmayer from my side this one is ready for review :)

**Comment by PSchmiedmayer** [comment]
Thank you @NikolaiMadlener! I will take a look at the PR later this week.

I have also approved it to run on our build infrastructure. It would be great if you can inspect the output of the builds and the resulting test coverage and see if you can add a small UI test to test that the rough functionality is working. No need to check all the dependencies in the UI itself. Please also take a look at the failing SwiftLint rule.

Regarding the build error: I suspect that we need to add the flag described here in our Fastlane Xcodebuild setup to ensure that the package can be run in the CI environment: https://forums.swift.org/t/telling-xcode-14-beta-4-to-trust-build-tool-plugins-programatically/59305

**Comment by NikolaiMadlener** [comment]
Sorry for spamming the CI, had some problems with SwiftLint which didn't warn me about the Accessibility Labels locally 🤔 Added them where I could, and added SwiftLint ignore statements for now where I couldn't (Spezi Framework files).

Just saw the #25 issue above, I could take that over next if that's fine.

**Comment by philippzagar** [comment]
@NikolaiMadlener Nice stuff, love the dynamic fetching of packages via the `PackageHelper`! 🚀

File: TemplateApplication/Contributions/ContributionsList.swift [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+
+struct ContributionsList: View {
+    var packages: [Package] = PackageHelper.getPackageList()
+    
+    var body: some View {
+        NavigationView {
+            List {
+                Section(footer: Text("This project is licensed under the MIT License.")) {
+                    Text("The following list contains all package dependencies SpeziTemplateApplication relies on.")
```
Comment (PSchmiedmayer) [inline TemplateApplication/Contributions/ContributionsList.swift:20]:
```suggestion
                    Text("The following list contains all Swift Package dependencies of the SpeziTemplateApplication.")
```

File: TemplateApplication/Contributions/ContributionsList.swift [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                }
+                Section(header: Text("Packages"), footer: Text("For packages without license labels refer to the individual repository links.")) {
```
Comment (PSchmiedmayer) [inline TemplateApplication/Contributions/ContributionsList.swift:22]:
```suggestion
                Section(header: Text("Packages"), footer: Text("Please refer to the individual repository links for packages without license labels.")) {
```

**Review by PSchmiedmayer — CHANGES_REQUESTED** [review]
@NikolaiMadlener Thank you for the great improvements and additions to the template application! I really like to approach using `SwiftPackageList`, which is a very smooth integration in the build pipeline.

Also, it's great to see the addition to SpeziAccount that will then embed this information in the Account overview; this should then replace the `ZStack` that you have in place here. 

Before doing a more detailed review when these elements are in place, I can already suggest the following:
1. The current implementation is based on a lot of the foundational work of the `SwiftPackageList` package and the views and examples from this package. I would suggest that you clearly highlight the code that you have reused from the project in the license identifier and properly provide attribution. Also, make sure that you are allowed to reuse the code here according to the license of the package.
2. The current implementation does not use the Xcode/Swift localization mechanism. I have recently improved the setup in #39, and this might directly benefit this PR. I would suggest that we localize all the values here as well.
3. I would suggest merging the latest version of the main branch into your fork + feature branch to ensure that the latest version of the CI is used and that you can automatically check the builds in this PR.
4. It would be great if we could support all the different license types that we currently have in our dependencies to complete the view.

And not sure if this is a misinterpretation, bit the version numbers seem to be slightly indented compared to the name:
Screenshot 2023-10-04 at 1 30 20 AM

**Comment by NikolaiMadlener** [comment]
Thanks for the great first feedback @PSchmiedmayer ! Will adopt all of the points.
Exactly, ZStack will be removed once my PR with the addition to SpeziAccount is merged.

**Comment by PSchmiedmayer** [comment]
Sounds great; thank you @NikolaiMadlener 👍

**Comment by NikolaiMadlener** [comment]
> @NikolaiMadlener Thank you for the great improvements and additions to the template application! I really like to approach using `SwiftPackageList`, which is a very smooth integration in the build pipeline.
> 
> Also, it's great to see the addition to SpeziAccount that will then embed this information in the Account overview; this should then replace the `ZStack` that you have in place here.
> 
> Before doing a more detailed review when these elements are in place, I can already suggest the following:
> 
> 1. The current implementation is based on a lot of the foundational work of the `SwiftPackageList` package and the views and examples from this package. I would suggest that you clearly highlight the code that you have reused from the project in the license identifier and properly provide attribution. Also, make sure that you are allowed to reuse the code here according to the license of the package.
> 2. The current implementation does not use the Xcode/Swift localization mechanism. I have recently improved the setup in [Use String Catalogues & Update to the Latest Spezi Packages #39](https://github.com/StanfordSpezi/SpeziTemplateApplication/pull/39), and this might directly benefit this PR. I would suggest that we localize all the values here as well.
> 3. I would suggest merging the latest version of the main branch into your fork + feature branch to ensure that the latest version of the CI is used and that you can automatically check the builds in this PR.
> 4. It would be great if we could support all the different license types that we currently have in our dependencies to complete the view.
> 
> And not sure if this is a misinterpretation, bit the version numbers seem to be slightly indented compared to the name: Screenshot 2023-10-04 at 1 30 20 AM

@PSchmiedmayer Your initial feedback should be implemented now. Regarding 1., `SwiftPackageList` is licensed under MIT so the usage should be fine. Also attributed the code that is based on the work of the `SwiftPackageList` repo now (as a comment within the `Package+LicenseType.swift`). 
Please let me know if there is anything else I can improve before this can be merged.

**Pull request merged**